cmd/state-svc/internal/hash/file_hasher.go

```diff
@@ -53,47 +56,61 @@ func (fh *FileHasher) HashFiles(wd string, globs []string) (_ string, _ []hashed
 				}
 				f = af
 			}
-			file, err := os.Open(f)
+			fileInfo, err := os.Stat(f)
 			if err != nil {
-				return "", nil, errs.Wrap(err, "Could not open file: %s", file.Name())
+				return "", nil, errs.Wrap(err, "Could not stat file: %s", f)
 			}
-			defer rtutils.Closer(file.Close, &rerr)
 
-			fileInfo, err := file.Stat()
-			if err != nil {
-				return "", nil, errs.Wrap(err, "Could not stat file: %s", file.Name())
+			if fileInfo.IsDir() {
+				continue
 			}
 
 			var hash string
-			cachedHash, ok := fh.cache.Get(cacheKey(file.Name(), fileInfo.ModTime()))
+			cachedHash, ok := fh.cache.Get(cacheKey(fileInfo.Name(), fileInfo.ModTime()))
 			if ok {
 				hash, ok = cachedHash.(string)
 				if !ok {
 					return "", nil, errs.New("Could not convert cache value to string")
 				}
 			} else {
 				fileHasher := xxhash.New()
+				// include filepath in hash, because moving files should affect the hash
+				fmt.Fprintf(fileHasher, "%016x", f)
+				file, err := os.Open(f)
+				if err != nil {
+					return "", nil, errs.Wrap(err, "Could not open file: %s", f)
+				}
+				defer file.Close()
 				if _, err := io.Copy(fileHasher, file); err != nil {
-					return "", nil, errs.Wrap(err, "Could not hash file: %s", file.Name())
+					return "", nil, errs.Wrap(err, "Could not hash file: %s", fileInfo.Name())
 				}
 
 				hash = fmt.Sprintf("%016x", fileHasher.Sum64())
 			}
 
-			fh.cache.Set(cacheKey(file.Name(), fileInfo.ModTime()), hash, cache.NoExpiration)
+			fh.cache.Set(cacheKey(fileInfo.Name(), fileInfo.ModTime()), hash, cache.NoExpiration)
 
+			hashes = append(hashes, hash)
 			hashedFiles = append(hashedFiles, hashedFile{
 				Pattern: glob,
-				Path:    file.Name(),
+				Path:    f,
 				Hash:    hash,
 			})
-
-			// Incorporate the individual file hash into the overall hash in hex format
-			fmt.Fprintf(hasher, "%016x", hash)
 		}
 	}
 
-	return fmt.Sprintf("%016x", hasher.Sum64()), hashedFiles, nil
+	if hashedFiles == nil {
+		return "", nil, nil
+	}
+
+	// Ensure the overall hash is consistently calculated
+	sort.Slice(hashedFiles, func(i, j int) bool { return hashedFiles[i].Path < hashedFiles[j].Path })
+	h := xxhash.New()
+	for _, f := range hashedFiles {
+		fmt.Fprintf(h, "%016x", f.Hash)
+	}
```

Comment:
Changed it to calculate the overall hash based on a sorted slice of all the individual hashes for all files. This way we get the same hash even if the files are sorted differently, or more importantly (what I ran into); if the patterns are different but produce the same result.